### ISSUE TITLE
Error indicates lifetime ends too early, when in fact it starts too late.

### ISSUE THREAD

**Issue by joshua-maros** [rust-lang/rust#117109]
This is not exactly a bug, just a misleading error.

Consider this invalid code:

```rust
struct BindGroup;
struct RenderPass<'a>(&'a mut ());

impl Drop for BindGroup {
    fn drop(&mut self) {}
}

impl<'a> Drop for RenderPass<'a> {
    fn drop(&mut self) {}
}

fn test<'rp, 'bg: 'rp>(holder: &mut RenderPass<'rp>, other: &'bg BindGroup) {
    todo!();
}


fn main() {
    let mut void = ();
    // Putting let bind_group = ... before let mut render_pass = ... fixes the code.
    let mut render_pass = RenderPass(&mut void);
    let bind_group = BindGroup;
    test(&mut render_pass, &bind_group);
    drop(render_pass);
    drop(bind_group);
}
```

The error message is as follows:

```
error[E0597]: `bind_group` does not live long enough
  --> src/main.rs:21:28
   |
20 |     let bind_group = BindGroup;
   |         ---------- binding `bind_group` declared here
21 |     test(&mut render_pass, &bind_group);
   |                            ^^^^^^^^^^^ borrowed value does not live long enough
...
24 | }
   | -
   | |
   | `bind_group` dropped here while still borrowed
   | borrow might be used here, when `render_pass` is dropped and runs the `Drop` code for type `RenderPass`
   |
   = note: values in a scope are dropped in the opposite order they are defined
```

However, it is impossible for `render_pass` to be dropped at the end of the scope as it is manually dropped earlier. Indeed, adding another `drop(render_pass)` at the end of the scope will cause another error indicating that `render_pass` has already been moved. 

Instead, the issue is that `bind_group` was created *after* render_pass. This causes the lifetime to be insufficient as it starts later than the lifetime of `render_pass`, making it not bigger than that lifetime. It would be nice if an error message could indicate this, as I had to spend significant time paring down my code to find the actual source of the issue.

This happens on stable and nightly according to [the playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=e698c88179ed68f589b9d304226cd421)

### Meta

`rustc --version --verbose`:
```
rustc 1.75.0-nightly (1c05d50c8 2023-10-21)
binary: rustc
commit-hash: 1c05d50c8403c56d9a8b6fb871f15aaa26fb5d07
commit-date: 2023-10-21
host: x86_64-unknown-linux-gnu
release: 1.75.0-nightly
LLVM version: 17.0.3
```

**Comment by RodBurman** [comment]
The fact the note says: "values in a scope are dropped in the opposite order they are defined", indicates the actual problem has been detected by the compiler. So it might be a simple fix to append this note to the current error or use it to modify the text as in the original suggestion, the intent of which I certainly agree with. Note the problem persists in Rust version 1.81.0.
